File: apps/api/app/schemas/meetings.py
"""Pydantic meeting schemas — the wire contract, mirrored in packages/shared."""

from datetime import date
from typing import Literal

from pydantic import BaseModel

from .items import ActionItem, Priority


class ExtractedItem(BaseModel):
    """One extractor item — no id yet; '' means none (NULL in the repository)."""

    title: str
    owner: str
    priority: Priority
    # "YYYY-MM-DD" parses to a date; "" means none; garbage 422s here.
    due: date | Literal[""]
    note: str


class Meeting(BaseModel):
    """One captured meeting — mirrors packages/shared. itemCount is derived."""

    id: int
    title: str
    capturedAt: str
    itemCount: int


class CreateMeetingRequest(BaseModel):
    """POST /api/meetings body: persist a capture and its extracted items."""

    title: str
    rawNotes: str
    items: list[ExtractedItem]


class CreateMeetingResponse(BaseModel):
    """POST /api/meetings response: the created meeting and its items."""

    meeting: Meeting
    items: list[ActionItem]


class MeetingDetail(BaseModel):
    """GET /api/meetings/{id} — one full capture: transcript plus its items."""

    id: int
    title: str
    rawNotes: str
    capturedAt: str
    itemCount: int
    items: list[ActionItem]


class MeetingsPage(BaseModel):
    """GET /api/meetings — one keyset page, newest first; None cursor = no more."""

    meetings: list[Meeting]
    nextCursor: str | None

File: packages/shared/src/meetings.ts
/** Contract for /api/meetings — a "capture" is a saved meeting (raw notes +
 * extracted items). Mirrored by hand in app/schemas/meetings.py. */

import { z } from "zod";

import { ActionItem } from "./items";
import { ExtractedItem } from "./extraction";

/** One captured meeting, as listed on Capture's RECENT strip. */
export const Meeting = z.object({
  id: z.number(),
  title: z.string(),
  /** ISO-8601 timestamp, stamped by the server at capture time. */
  capturedAt: z.string(),
  /** Derived (COUNT of the meeting's items) — never stored. */
  itemCount: z.number(),
});
export type Meeting = z.infer<typeof Meeting>;

/** POST /api/meetings request body — persist a capture and its extracted items. */
export const CreateMeetingRequest = z.object({
  title: z.string(),
  rawNotes: z.string(),
  items: z.array(ExtractedItem),
});
export type CreateMeetingRequest = z.infer<typeof CreateMeetingRequest>;

/** POST /api/meetings response body: the created meeting and its persisted items. */
export const CreateMeetingResponse = z.object({
  meeting: Meeting,
  items: z.array(ActionItem),
});
export type CreateMeetingResponse = z.infer<typeof CreateMeetingResponse>;

/** GET /api/meetings with `cursor` — one keyset page, newest first. */
export const MeetingsPage = z.object({
  meetings: z.array(Meeting),
  nextCursor: z.string().nullable(),
});
export type MeetingsPage = z.infer<typeof MeetingsPage>;

/** GET /api/meetings/{id} — a Meeting plus its transcript and the items
 * extracted from it, said as the relationship (extend), not a restated shape. */
export const MeetingDetail = Meeting.extend({
  rawNotes: z.string(),
  items: z.array(ActionItem),
});
export type MeetingDetail = z.infer<typeof MeetingDetail>;
